src/charm.py

```diff
@@ -0,0 +1,222 @@
+
+        # don't complete install until passwords set
+        if not self.state.peer_relation:
+            self.unit.status = WaitingStatus("waiting for peer relation")
+            event.defer()
+            return
+
+        if self.unit.is_leader() and not self.state.cluster.internal_user_credentials:
+            for user in CHARM_USERS:
+                self.state.cluster.update({f"{user}-password": self.workload.generate_password()})
+
+    def _on_cluster_relation_changed(self, event: EventBase) -> None:
+        """Generic handler for all 'something changed, update' events across all relations."""
+        # not all methods called
+        if not self.state.peer_relation:
+            return
+
+        # attempt startup of server
+        if not self.state.unit_server.started:
+            self.init_server()
+
+        # don't delay scale-down leader ops by restarting dying unit
+        if getattr(event, "departing_unit", None) == self.unit:
+            return
+
+        if (
+            self.config_manager.config_changed()
+            and self.state.unit_server.started
+            # and self.upgrade_events.idle
+        ):
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
+
+    def _on_secret_changed(self, event: SecretChangedEvent):
+        """Reconfigure services on a secret changed event."""
+        if not event.secret.label:
+            return
+
+        if not self.state.cluster.relation:
+            return
+
+        if event.secret.label == self.state.cluster.data_interface._generate_secret_label(
+            PEER,
+            self.state.cluster.relation.id,
+            None,  # type:ignore noqa
+        ):  # Changes with the soon upcoming new version of DP-libs STILL within this POC
+
+            if (
+                self.config_manager.config_changed()
+                and self.state.unit_server.started
+                # and self.upgrade_events.idle
+            ):
+                logger.info(f"Secret {event.secret.label} changed.")
+
+    def _start(self, event: EventBase) -> None:
+        """Forces a rolling-restart event.
+
+        Necessary for ensuring that `on_start` restarts roll.
+        """
+        # if not self.state.peer_relation or not self.state.stable or not self.upgrade_events.idle:
+        self.unit.status = MaintenanceStatus("Starting...")
+        if not self.state.peer_relation or not self.state.stable:
+            event.defer()
+            return
+
+        # not needed during application init
+        # only needed for scenarios where the LXD goes down (e.g PC shutdown)
+        if not self.workload.alive():
+            self.on[f"{self.restart.name}"].acquire_lock.emit()
+        else:
+            self.unit.status = ActiveStatus()
+
+    def _restart(self, event: EventBase) -> None:
+        """Handler for emitted restart events."""
+        # if not self.state.stable or not self.upgrade_events.idle:
+        if not self.state.stable:
+            event.defer()
```

Comment:
This defer will cause you a problem: https://github.com/canonical/charm-rolling-ops/issues/11

Reply:
Hmmm, thank you... In fact this check is not that important for us (it was for ZK, where I "stole" it from, but not for Dashboards). So I may just remove the 2 lines above. Thx much

Reply:
Commented out for now, to be re-evaluated when adding upgrades